### PR TITLE
Fix reconnection issues

### PR DESCRIPTION
**Pull request by andrewvc**
Fix a bug where this plugin would not properly reconnect if it lost its connection to the server while the connection would re-establish itself, the queue subscription would not

/cc @GlenRSmith

### CODE REVIEW

**Comment by andrewvc** [comment]
I should add that I tested this manually as well, stopping and starting a rabbitmq server, it seemed to resubscribe just fine.

File: lib/logstash/inputs/rabbitmq.rb [hunk 1]
```diff
@@ -52,6 +52,10 @@ class RabbitMQ < LogStash::Inputs::Threadable
       # * Wildcards are not valid on direct exchanges.
       config :key, :validate => :string, :default => "logstash"
 
+      # Amount of time in seconds to wait after a failed subscription request
+      # before retrying. Subscribes can fail if the server goes away and then comes back
+      config :subscription_retry_interval_seconds, :validate => :number, :required => true, :default => 5
```
Comment (suyograo) [inline lib/logstash/inputs/rabbitmq.rb]: nice `_seconds` :)

Comment (suyograo) [inline lib/logstash/inputs/rabbitmq.rb]: why is this required, if it has a default?

**Comment by GlenRSmith** [comment]
Thanks, @andrewvc !

File: lib/logstash/inputs/rabbitmq.rb (hunk 1 above)
Comment (andrewvc) [inline lib/logstash/inputs/rabbitmq.rb]: In case we ever allow people to set a `null` value. Future proofing :) ? :) :smiling_imp:

File: lib/logstash/inputs/rabbitmq.rb [hunk 2]
```diff
@@ -93,15 +97,38 @@ def declare_queue
...
         # we manually build a consumer here to be able to keep a reference to it
         # in an @ivar even though we use a blocking version of HB::Queue#subscribe
-        @consumer = @hare_info.queue.build_consumer(:block => true) do |metadata, data|
-          @codec.decode(data) do |event|
-            decorate(event)
-            @output_queue << event if event
+
+        # The logic here around resubscription might seem strange, but its predicated on the fact
+        # that we rely on MarchHare to do the reconnection for us with auto_reconnect.
+        # Unfortunately, while MarchHare does the reconnection work it won't re-subscribe the consumer
```
Comment (michaelklishin) [inline lib/logstash/inputs/rabbitmq.rb]: Yes, March Hare will recover consumers. If it does not then it is a bug.

File: lib/logstash/inputs/rabbitmq.rb [hunk 3]
```diff
@@ -93,15 +97,38 @@ def declare_queue
...
+              @logger.info("Will subscribe with consumer", :config => config)
+              @hare_info.queue.subscribe_with(@consumer, :manual_ack => @ack, :block => true)
+              @logger.warn("Queue subscription ended! Will retry in #{@subscription_retry_interval_seconds}s", :config => config)
+            else
+              @logger.info("RabbitMQ connection down, will wait to retry subscription", :config => config)
+            end
+
+            break if stop?
+            @consumer.gracefully_shut_down # Try to clean up gracefully
+          rescue MarchHare::Exception => e
```
Comment (suyograo) [inline lib/logstash/inputs/rabbitmq.rb]: hmmm..Is there a specific exception that gets thrown? or a blanket `MarchHare::Exception`

File: lib/logstash/inputs/rabbitmq.rb (hunk 2 above)
Comment (andrewvc) [inline lib/logstash/inputs/rabbitmq.rb]: Hmmm, then I believe we have a bug!

Comment (andrewvc) [inline lib/logstash/inputs/rabbitmq.rb]: @michaelklishin if you run logstash with this config: https://github.com/logstash-plugins/logstash-input-rabbitmq/blob/master/examples/rabbitmq_stdout.conf and restart rabbitmq while logstash is running logstash will exit because the consumer will stop blocking and exit.

File: lib/logstash/inputs/rabbitmq.rb (hunk 3 above)
Comment (andrewvc) [inline lib/logstash/inputs/rabbitmq.rb]: We have to leave it generic esp. because of future exceptions here.

**Comment by suyograo** [comment]
LGTM. When the upstream fix gets in, we can revert this logic

File: lib/logstash/inputs/rabbitmq.rb (hunk 2 above)
Comment (michaelklishin) [inline lib/logstash/inputs/rabbitmq.rb]: Blocking consumers are expected to unblock before recovery. Using a blocking consumer makes certain sense in small scripts but I see no justification for doing it in Logstash.
